### PR TITLE
[MXNET-867] Pooling1D with "same" padding

### PR DESCRIPTION
**Pull request by ChaiBapchya**
## Description ##
MaxPooling 1D added another pooling convention (aka padding type) called "same"

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by stu1130** [comment]
Thanks for your contribution @ChaiBapchya 
@mxnet-label-bot[pr-awaiting-review]

**Comment by stu1130** [comment]
Thanks for your contribution @ChaiBapchya 
@mxnet-label-bot[pr-awaiting-review]

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -6957,6 +6957,17 @@ def test_invalid_depth_dim():
     test_invalid_block_size()
     test_invalid_depth_dim()
 
+@with_seed()
+def test_max_pooling_pad_type_same():
+    input_data=mx.nd.array(np.random.rand(1,1,10))
+    output_data=mx.nd.Pooling(
+        input_data,
+        kernel=(2),
```
Comment (anirudh2290) [inline tests/python/unittest/test_operator.py]: what about other kernel.ndim() >= 2

Comment (ChaiBapchya) [reply tests/python/unittest/test_operator.py]: Specifically address the JIRA ticket that asks for maxpool1D

Comment (anirudh2290) [reply tests/python/unittest/test_operator.py]: so we should check and raise an exception if it is used with other ndims

File: src/operator/nn/pooling.cc [hunk 2]
```diff
@@ -126,8 +126,8 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
       oshape[2] = 1 +
                   (dshape[2] + 2 * param.pad[0] - param.kernel[0]) /
                       param.stride[0];
-    } else {
-      oshape[2] = 1 + static_cast<int>(std::ceil(
+    } else if (param.pooling_convention == pool_enum::kFull) {
```
Comment (access2rohit) [inline src/operator/nn/pooling.cc]: You introduced "else if" and removed "else". So there is no default behavior here. Are you sure all the cases will fall under these 2 conditionals only ?

File: tests/python/unittest/test_operator.py (hunk 1 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]: @ChaiBapchya You should add checks for illegal inputs that are not supported by your implementation.

Comment (ChaiBapchya) [reply tests/python/unittest/test_operator.py]: Yup, I'm working on it.

File: src/operator/nn/pooling.cc [hunk 3]
```diff
@@ -96,6 +96,11 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
     CHECK(param.p_value.has_value());
   }
   const TShape &dshape = (*in_shape)[0];
+  if (param.pooling_convention == pool_enum::kSame) {
+    CHECK_EQ(dshape.ndim(), 3U)
+      << "Pooling: Input data should be 3D in (batch, channel, x)"
+      << ", Currently 'same' supports Max Pooling 1-D";
```
Comment (apeforest) [inline src/operator/nn/pooling.cc]: nit: change coma to period.

File: src/operator/nn/pooling.cc [hunk 4]
```diff
@@ -126,11 +131,15 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
...
-    } else {
-      oshape[2] = 1 + static_cast<int>(std::ceil(
+    } else if (param.pooling_convention == pool_enum::kFull) {
+      oshape[2] = 1 + static_cast<int>(ceil(
                           static_cast<float>(dshape[2] + 2 * param.pad[0] -
                                              param.kernel[0]) /
                           param.stride[0]));
+    } else {
+      oshape[2] = static_cast<int>(ceil(
+                          static_cast<float>(dshape[2] + 2 * param.pad[0]) /
```
Comment (apeforest) [inline src/operator/nn/pooling.cc]: nit: indentation

File: src/operator/nn/pooling.cc (hunk 3 above)
Comment (ChaiBapchya) [reply src/operator/nn/pooling.cc]: My bad. Done

File: tests/python/unittest/test_operator.py [hunk 5]
```diff
@@ -6957,6 +6957,21 @@ def test_invalid_depth_dim():
...
     test_invalid_depth_dim()
 
+@with_seed()
+def test_max_pooling_pad_type_same():
+    import math
+    input_data = mx.nd.array(np.random.rand(1,1,10))
+    stride = 2
+    kernel = 2
+    output_data=mx.nd.Pooling(
+        input_data,
```
Comment (apeforest) [inline tests/python/unittest/test_operator.py]: Do you also want to test the case where users selects "same" pad type but also supplies pad values?

**Comment by apeforest** [comment]
Thanks for making this contribution. LGTM overall. You may also want to update the operator description in operator/nn/pooling.cc since you added a new type "same"

File: src/operator/nn/pooling.cc (hunk 4 above)
Comment (ChaiBapchya) [reply src/operator/nn/pooling.cc]: Unable to find the indentation issue. Also is there a similar linting tool for python to ensure I don't miss out on those. Do we use pylint?

Comment (apeforest) [reply src/operator/nn/pooling.cc]: If pylint does not complain, I guess it's fine. But it seems not following the Google python style guide: https://github.com/google/styleguide/blob/gh-pages/pyguide.md

**Comment by ChaiBapchya** [comment]
As discussed with Sina Afrooze and Lin, this is just the first step towards complete solution/fix.

Immediate demand was for Max Pool 1 Dimension to support "same" padding (the way Tensorflow does). Above pull request addresses that.

However, for a complete solution, following things need to be understood

### Asymmetric Padding
Feature request is about "Asymmetric padding"
Currently, MXNet supports Symmetric padding
e.g. Input data -> `1,2,3,4` and `pad`=`2` 
Output data -> `0,0,1,2,3,4,0,0`
 
#### Why isn't Asymmetric padding supported by CUDnn, MXNet?

Due to misalignment of input-output if padding happens only on one side (left or right alone).
It is called as *"Causal Convolution"* (only one side pad)

### Current work-around
Level of abstraction introduced.
By default, one can use `pad` parameter while performing `Pooling` or `Convolution` which performs symmetric padding. But, if there is a specific use-case for performing asymmetric padding, one has to perform following steps,
Perform independent padding (using `pad` operator) and then perform `Pooling` or `Convolution` as desired.

### Future steps
If one has to fully implement asymmetric padding, following changes need to be made
1. Pooling
- Add another parameter such as pad_type (because currently, `pooling_conv` which allows - 'full', 'valid', 'same' has been misused. Original meaning of `pooling_conv` was not supposed to take care of "type of padding"

2. Convolution
- Incorporate asymmetric padding for the same

**Comment by anirudhacharya** [comment]
@ChaiBapchya can you create and issue for the future steps, giving the full description as in the above comment

**Comment by ChaiBapchya** [comment]
Sure. Thanks for suggesting :)

**Pull request merged**